# 11.12 오류 코드를 예외로 바꾸기
---

```jsx
if(data)
	return new ShippingRules(data);
else
	return -23;
```

```jsx
if(data)
	return new ShippingRules(data);
else
	throw new OrderProcessingError(-23);
```

# 언제 오류코드를 사용하고 언제 예외를 사용해야 하나

예외는 `정확히` 예상 밖의 동작일 때만 쓰여야 한다. 프로그램의 정상 동작 범주에 들지 않는 오류를 나타낼 때만 써야한다. 이를 판단하는 기준은 예외를 던지는 코드를 프로그램 종료 코드로 바꿔도 프로그램이 여전히 정상 동작할지를 따져보는 것이다. 정상 동작하지 않을 것 같다면 예외 대신 오류를 검출하여 프로그램을 정상 흐름으로 되돌리게끔 처리해야한다.

# 예외처리 예시

아래는 배송지의 배송 규칙을 알아내는 코드이다.

```jsx
function localShippingRules(country){
	const data = countryData.shippingRules[country];
	if (data) return new ShippingRules(data);
	else return -23;
}
```

위 코드에서 country가 유효한지는 이 함수 호출 전에 다 검증했다고 가정한다.

따라서 이 함수에서 오류가 난다면 무언가 문제가 생겼다는 것을 뜻한다.

```jsx
function calculateShippingCosts(anOrder){
	//....
	const shippingRules = localShippingRules(anOrder.country);
	if(shippingRules < 0) return shippingRules; //오류 검증
	//....
}
```

```jsx
const status = calculateShippingCosts(orderData);
if (status < 0) errorList.push({order: orderData, errorCode: status});
```

오류를 낸 주문 목록을 errorList에 넣는다.

현재 위 로직은 예외처리가 아닌 오류 검증을 하는 방식으로 돼있다. 여기서 위 로직을 예외를 사용하는 로직으로 변경하기 위해서는 다음과 같은 질문을 던져봐야 한다.

- localShippingRules()는 배송 규칙들이 countryData에 제대로 반영되어 있는지
- country 인수가 전역 데이터에 저장된 키들과 일치하는 곳에서 가져온 것인지

위 질문들의 답이 긍정적이라면 오류코드를 예외로 바꿔도 된다.

## 예외코드로 변경하기

1. 최상위에 예외 핸들러를 갖춘다. 

```jsx
try{
	const status = calculateShippingCosts(orderData);
} catch (e){
	//예외 처리 로직
}
if (status < 0) errorList.push({order: orderData, errorCode: status});
```

하지만 이렇게 하면 status의 유효범위가 try 블록으로 제한되어 조건문에서 검사할 수 없기 때문에 위와 같이 할 수는 없다. 그래서 status 선언과 초기화를 분리해야한다.

```jsx
let status;
try{
	status = calculateShippingCosts(orderData);
} catch (e){
	throw e;
}
if (status < 0) errorList.push({order: orderData, errorCode: status});
```

1. 이번 리팩터링으로 추가된 예외만을 처리하고자 한다면 다른 예외와 구별 방법이 필요하다. 이 경우 서브클래스를 만드는게 가장 자연스럽다. 자바스크립트는 여기 해당하지 않지만 이 책의 저자는 다음과 같은 방법으로 구현하는 것을 좋아한다고 한다.

```jsx
class OrderProcessingError extends Error{
	constructor(errorCode){
		super(`주문 처리 오류: ${errorCode}`);
		this.code = errorCode;	
	}
	get name() {return "OrderProcessingError";}
}
```

1. 위 클래스가 준비되면 오류 코드를 처리할 때와 같은 방식으로 이 예외 클래스를 처리하는 로직을 추가할 수 있다.

```jsx
let status;
try{
	status = calculateShippingCosts(orderData);
} catch (e){
	if (e instanceof OrderProcessingError)
		errorList.push({order: orderData, errorCode: e.code});
	else
		throw e;
}
if (status < 0) errorList.push({order: orderData, errorCode: status});
```

1. `localShippingRules` 에서 오류 대신 예외를 던지도록 변경한다.

```jsx
function localShippingRules(country){
	const data = countryData.shippingRules[country];
	if (data) return new ShippingRules(data);
	else throw new OrderProcessingError(-23);
}
```

1. 아래와 같은 코드를 작성해서 최종적으로 테스트한다.

```jsx
function calculateShippingCosts(anOrder){
	//....
	const shippingRules = localShippingRules(anOrder.country);
	if(shippingRules < 0) throw new Error("오류 코드가 다 사라지지 않았습니다.");
	//....
}
```

1. 위의 단계까지 확인했다면 이제 status 변수를 제거한다

```jsx
try{
	calculateShippingCosts(orderData);
} catch (e){
	if (e instanceof OrderProcessingError)
		errorList.push({order: orderData, errorCode: e.code});
	else
		throw e;
}
```
